packages/components/src/ui/icon/component.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * External dependencies
+ */
+import { contextConnect, useContextSystem } from '@wp-g2/context';
+import { css, cx } from '@wp-g2/styles';
+
+/**
+ * WordPress dependencies
+ */
+import { useMemo, cloneElement } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { View } from '../view';
+import * as styles from './styles';
+
+/* eslint-disable jsdoc/valid-types */
+/**
+ * @typedef Props
+ * @property {import('react').CSSProperties['fill']} [fill='currentColor'] Icon fill.
+ * @property {number} [height] Icon height.
+ * @property {import('react').ReactElement} icon Icon to render.
+ * @property {boolean} [inline] Whether the icon is inline with text.
+ * @property {number | string} [size=20] Icon size.
+ * @property {keyof styles} [variant] Variant to render.
+ * @property {number} [width] Icon width.
+ * @property {never} [children] Children are disallowed.
+ */
+/* eslint-enable jsdoc/valid-types */
+
+/**
+ * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'div'>} props
+ * @param {import('react').Ref<any>} forwardedRef
+ */
+function Icon( props, forwardedRef ) {
```

Comment:
The "official" icon component now, is actually in `@wordpress/icons` and the idea was to deprecate the one in "components" at some point, this feels like it goes in the opposite direction, so we should give it some thoughts.

Reply:
in other words, why do we need an "Icon" component in "components" and why the one in `@wordpress/icons` is not sufficient?

Reply:
> ...why the one in @wordpress/icons is not sufficient?

@youknowriad Good question! The short answer is custom styling (or rather refinements).

From my experience, SVG-based icons usually require adjustments based on width, height, display (block or inline), and color (fill or path). 

The current `@wordpress/icon` supports width and height, but are missing supprt for display + color.

To support this, custom styles have been added for the (Icon) `svg` selector, and is scattered throughout the `.scss` codebase:
https://github.com/WordPress/gutenberg/search?l=SCSS&p=4&q=svg

(Note: Most are color based)

The one from this PR addresses `display`, `color`, and is rendered with a wrapping `<View />`.
This wrapper helps control the layout rendering of the `svg` element, and offers improved style customization + stability (with the new styling system).

Hope this helps!

Reply:
The Icon component in the `@wordpress/icons` doesn't need to support color or display because the color is just the one that is actually just the `current-color` in CSS and display is more for its parent than the icon itself.

So instead of adding a 4th icon block and continue the fragmentation, I personally don't think it's worth it and we should just continue with the one we have.

Reply:
> the color is just the one that is actually just the current-color in CSS

The tricky thing is the current `@wordpress/icons` doesn't support this, so `fill: currentColor` CSS is added in places that need it. We can enhance `@wordpress/icons` to have this style be default - although I'd be uncertain how it should be done (either with a new `.scss` file or inline styles).

> display is more for its parent than the icon itself
That's true in some sense. But I think it's a common enough setup that it would be useful if supported by the component.

---

Taking a step back, the purpose of the component is to provide a nice "wrapper" for the `svg` icon.
Something that takes care of the stuff I had mentioned.

The best parallel example I can think of is how Github handles icons.

[Octicons](https://primer.style/octicons/) is their SVG icon library (https://primer.style/octicons/).
For their React component system, they have [`StyledOcticon`](https://primer.style/components/StyledOcticon), which is their "wrapper".

They do have a dedicated Icon renderer component within `Octicon`
https://github.com/primer/octicons/blob/74b249595db151ec361bcacfdca18215ae8cc020/docs/src/components/icon.js

(which interestingly has some layout opinions hard coded, like `display: inline-block` and `vertical-align: middle`)

---

We could have a similar setup where we leave `@wordpress/icon` as is, and we create a styled wrapper.
However, that touches upon the concern you mentioned...

> So instead of adding a 4th icon block

TLDR:

We need a reliable way to render icons within the new Component System.
The out-of-the-box `<Icon />` component from `@wordpress/icon` is missing a couple of things.
We can create a styled wrapper - something that's used internally, and only at the framework level.

Reply:
I guess the main difference between the two approaches are is whether you do:

```
<View config={config}>
    <Icon />
</View>
```

or 

```
<Icon config={config} />
```

In these situations, I personally always lean towards 1 because I prefer composability over inheritance.

Reply:
@youknowriad 

> lean towards 1 because I prefer composability over inheritance.

Option 1 would be simpler. But that would mean that it's more verbose to write, as we'd always need to do something like this for every usage of icon:

```jsx
<StyledIcon><Icon {...} /></StyledIcon>
```

I'm okay with this, but I'd prefer something simpler (to write) :).

If we move forward with this option, would we restrict the "Icon wrapper" component to living in `/components/ui` for now?

Thanks!

Reply:
What we need a `StyledIcon` instead of just `View`?

Reply:
@youknowriad The icon wrapper would take care of `display`, `color`, and `user-select` out of the box for the inner `Icon`. `View` is capable of doing this, but it doesn't these styles by default.

The custom styles can be found here:
https://github.com/ItsJonQ/g2/blob/main/packages/components/src/Icon/Icon.styles.js

Reply:
I guess I don't really understand why not just move these styles to `Icon` component in `@wordpress/icons` if they are mandatory. 

I don't want to be a blocker, so feel free to move on if you think otherwise but for me there's a lot of unnecessary duplication in terms of dependencies and useless props/behaviors in these new components and my concern is making sure we catch these.

Reply:
> I guess I don't really understand why not just move these styles to Icon component in @wordpress/icons if they are mandatory.

I think this is what we should do, ultimately replacing the existing `Icon` component in `@wordpress/icons` with the new one.

@ItsJonQ We'll need to brainstorm some way of doing this given the dependency on G2. We cannot do it just with inline styles, unfortunately, which might have simplified things.

For context: I was confused about the status of the Icon component in `@wordpress/components` which is why the adapter behaves the way it does and why the new Icon component lives where it does now.

I do not think we should go backwards but I think we should enhance the `@wordpress/icons` component with the new features that Q introduced in the CSS-in-JS one.

Reply:
@youknowriad + @sarayourfriend Thank you for your thoughts 🙏 

> We'll need to brainstorm some way of doing this given the dependency on G2. We cannot do it just with inline styles, unfortunately, which might have simplified things.

A part of me wishes we could do it with inline styles 😛 .

---

Ultimately, we want to provide `Icon` with some styles to make it easier to work with and more reliable to render (style wise).

The 3 choices we have are:

1. Use inline styles
2. Use CSS (via `.scss`)
3. Use CSS-in-JS (Emotion or the new Style System)

**Inline styles** is the only option that does **not** require a workflow adjustment.
The trade-off is that customizing styles may be quite difficult.

**CSS** would require the workflow adjustment of stylesheet compile/export/import.
Not the worst thing in the world, as this is how we're currently setup.

**CSS-in-JS** would require adding a dependency to compile the styles.
This is the direction we're heading with the new Component System.

---

If we go wit the Style System CSS-in-JS solution, it would be good if we could create the packages. That way, the style system could be shared between `@wordpress/icons` and `@wordpress/components`.

Bringing over the style system was something briefly touch upon in this issue under "Backfilling"
https://github.com/WordPress/gutenberg/issues/28399

---

**For now**, we may be able to work around this for the component integration work. However, this is something that we'll need to solve.